main/src/main/scala/sbt/Load.scala

```diff
@@ -648,54 +676,66 @@ object Load {
-        case p: Plugins          => pluginSettings(p)
-        case p: AutoPlugins      => autoPluginSettings(p)
-        case q: Sequence         => (Seq.empty[Setting[_]] /: q.sequence) { (b, add) => b ++ expandSettings(add) }
+    log: Logger): Project =
+    timed(s"Load.resolveProject(${rawProject.id})", log) {
+      import AddSettings._
+      // 1. Apply all the project manipulations from .sbt files in order
+      val transformedProject =
+        timed(s"Load.resolveProject(${rawProject.id}): transformedProject", log) {
+          configFiles.flatMap(_.manipulations).foldLeft(rawProject) { (prev, t) =>
+            t(prev)
+          }
+        }
+      // 2. Discover all the autoplugins and contributed configurations.
+      val autoPlugins =
+        timed(s"Load.resolveProject(${rawProject.id}): autoPlugins", log) {
+          try loadedPlugins.detected.deducePluginsFromProject(transformedProject, log)
+          catch { case e: AutoPluginException => throw translateAutoPluginException(e, transformedProject) }
+        }
+      val autoConfigs = autoPlugins.flatMap(_.projectConfigurations)
+
+      // 3. Use AddSettings instance to order all Setting[_]s appropriately
+      val allSettings = {
+        // TODO - This mechanism of applying settings could be off... It's in two places now...
+        lazy val defaultSbtFiles = configurationSources(transformedProject.base)
+        // Grabs the plugin settings for old-style sbt plugins.
+        def pluginSettings(f: Plugins) =
+          timed(s"Load.resolveProject(${rawProject.id}): expandSettings(...): pluginSettings($f)", log) {
+            val included = loadedPlugins.detected.plugins.values.filter(f.include) // don't apply the filter to AutoPlugins, only Plugins
+            included.flatMap(p => p.settings.filter(isProjectThis) ++ p.projectSettings)
+          }
+        // Filter the AutoPlugin settings we included based on which ones are
+        // intended in the AddSettings.AutoPlugins filter.
+        def autoPluginSettings(f: AutoPlugins) =
+          timed(s"Load.resolveProject(${rawProject.id}): expandSettings(...): autoPluginSettings($f)", log) {
+            autoPlugins.filter(f.include).flatMap(_.projectSettings)
+          }
+        // Grab all the settigns we already loaded from sbt files
+        def settings(files: Seq[File]): Seq[Setting[_]] =
+          timed(s"Load.resolveProject(${rawProject.id}): expandSettings(...): settings($files)", log) {
+            for {
+              file <- files
+              config <- (memoSettings get file).toSeq
+              setting <- config.settings
+            } yield setting
+          }
+        // Expand the AddSettings instance into a real Seq[Setting[_]] we'll use on the project
+        def expandSettings(auto: AddSettings): Seq[Setting[_]] = auto match {
+          case BuildScalaFiles     => rawProject.settings
+          case User                => globalUserSettings.cachedProjectLoaded(loadedPlugins.loader)
```

Comment:
The correctness hinges on the fact that it will be ok. As I see it the only place where you see `projectLoaded` gets changed is in `loadGlobalSettings` (https://github.com/sbt/sbt/pull/2630/files/c67b3881de6287c7c970b1bf5a15c9c1739845ef#diff-1b5b6f2f3765a38e3ca6b130a15076ffL72). I am guessing that the output of `buildGlobalSettings` is stable across subprojects. Am I wrong here?